pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -1097,19 +1097,30 @@ func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1beta
 	}
 
 	if _, needsRemediation := machine.Annotations[externalRemediationAnnotation]; !needsRemediation {
-		return nil
-	}
-
-	node, err := a.getNodeByMachine(ctx, machine)
+		// It can happen that an unhealthy machine with external remediation annotation got healthy, and the external
+		// remediation annotation was removed by the MHC controller, before remediation finished or even started
+		// (e.g. the metal3 pod is not running, but machine was rebooted manually).
+		// In this case stop remediation, if possible.
+
+		// If bmh isn't requested to power off, and the machine isn't marked as powered off for remediation,
+		// there is no ongoing remediation, and there is nothing to do
+		powerOffRequestExists := hasPowerOffRequestAnnotation(baremetalhost)
+		_, poweredOffForRemediationExists := machine.Annotations[poweredOffForRemediation]
+		if !powerOffRequestExists && !poweredOffForRemediationExists {
```

Comment:
I think that `!poweredOffForRemediationExists` can be omitted. We do the same check in line 1116

Reply:
and then you'll be able to omit `powerOffRequestExists` condition from line 1116, since you got there only after you verified it's `true`

Reply:
> I think that !poweredOffForRemediationExists can be omitted

I think it is important to check this one: the powerOffRequest is removed before the end of the remediation, so poweredOffForRemediationExists is the one indicating a still ongoing remediation

Reply:
indeed it's important to check that it exists, but no need to check it twice :)

Reply:
At this place I want to check if we have an ongoing remediation. If not, we can just return as in the old version.
Also after looking at the latest version of the overall code (I was on an old version :/ ), I think we need to check both annotations: bmh powerOffRequest indicates remediation started (it is the first step), and machine poweredOffForRemediation indicates it is still ongoing (removing it is the last step). Only if both are false, we can be sure that there is no ongoing remediation and return. If we would only check powerOffRequest, we could be at a later stage of remediation, where the host was powered on again, but the node wasn't restored yet, so the remediation process has to continue. I updated the comment to explain this.